python/paddle/distributed/fleet/utils/hybrid_parallel_inference.py

```diff
@@ -0,0 +1,785 @@
+                layers.array_write(other_var, i=step_idx, array=arr)
+                
+                # update cond and assign to cond_int, we will sync cond_int
+                layers.assign(layers.cast(cond, dtype="int32"), cond_int)
+                
+            with paddle.fluid.device_guard(f'{model._device}:all'):
+                # the code below must at end of while block and exists in device:all
+                layers.assign(layers.cast(cond_int, dtype='bool'), cond)
+                
+        with paddle.fluid.device_guard(f'{model._device}:all'):
+            # use a empty lod_tensor_array to clear lod_tensor_array
+            layers.assign(layers.create_array(data.dtype), arr)
+    """
+
+    def __init__(self,
+                 startup_program,
+                 main_program,
+                 num_mp=1,
+                 num_pp=1,
+                 micro_batch_size=1,
+                 beam_size=1,
+                 init_comm=True,
+                 role_maker=None):
+
+        assert isinstance(startup_program, Program)
+        assert isinstance(main_program, Program)
+
+        self._device = None
+        if core.is_compiled_with_npu():
+            self._device = "npu"
+        elif core.is_compiled_with_cuda():
+            self._device = "gpu"
+        assert self._device, "Only gpu and npu are supported."
+        assert not in_dygraph_mode(), "Only static mode is supported."
+
+        op_maker = core.op_proto_and_checker_maker
+        self._op_role = op_maker.OpRole
+        self._op_role_key = op_maker.kOpRoleAttrName()
+        self._op_device_key = op_maker.kOpDeviceAttrName()
+
+        self._param_device_map = dict()
+
+        self._pipeline_pair = []
+        self._pipeline_pair_in_while = []
+        self._pp_ring_map = dict()
+        self.ring_id = 20  # Just a magic number
+
+        self.micro_batch_size = micro_batch_size
+        self.beam_size = beam_size
+        self.num_mp = num_mp
+        self.num_pp = num_pp
+        self.init_comm = init_comm
+
+        self._output_var_to_op = None
+        self._input_var_to_op = None
+        self._main_program = main_program
+        self._startup_program = startup_program
+
+        if role_maker is None:
+            self.role_maker = fleet.base.role_maker.PaddleCloudRoleMaker(
+                is_collective=True)
+        else:
+            if isinstance(role_maker, fleet.base.role_maker.RoleMakerBase):
+                assert role_maker._is_collective == True
+                self.role_maker = role_maker
+
+        # communication_group info
+        self.mp_ring_id = 0
+        self.global_ring_id = 1
+
+        self.endpoints = self.role_maker._get_trainer_endpoints()
+        self.current_endpoint = self.endpoints[self.role_maker._worker_index()]
+        self.rank = self.role_maker._worker_index()
+        self.nranks = self.role_maker._worker_num()
+        assert self.nranks % self.num_mp == 0
```

Comment:
I think num_mp and num_pp are not necessary, since the two values can be inferred from the program. For example, we can get num_pp by the number of devices used in device_guard and the 'num_mp = total_cards/num_pp'

Reply:
这个主要是用于校验，还是有必要的，参考176行